source/particle/property/cpo_bingham_average.cc

```diff
@@ -0,0 +1,369 @@
+        const std::array<std::pair<double,Tensor<1,3,double>>, 3> eigenvectors_b = eigenvectors(sum_matrix_b, SymmetricTensorEigenvectorMethod::jacobi);
+        const std::array<std::pair<double,Tensor<1,3,double>>, 3> eigenvectors_c = eigenvectors(sum_matrix_c, SymmetricTensorEigenvectorMethod::jacobi);
+
+
+        const Tensor<1,3,double> averaged_a = eigenvectors_a[0].second * eigenvectors_a[0].first;
+        const Tensor<1,3,double> averaged_b = eigenvectors_b[0].second * eigenvectors_b[0].first;
+        const Tensor<1,3,double> averaged_c = eigenvectors_c[0].second * eigenvectors_a[0].first;
+
+        return
+        {
+          {
+            {{averaged_a[0],averaged_a[1],averaged_a[2]}},
+            {{averaged_b[0],averaged_b[1],averaged_b[2]}},
+            {{averaged_c[0],averaged_c[1],averaged_c[2]}}
+          }
+        };
+      }
+
+      template<int dim>
+      std::vector<Tensor<2,3>>
+      CpoBinghamAverage<dim>::random_draw_volume_weighting(std::vector<double> fv,
+                                                           std::vector<Tensor<2,3>> matrices,
+                                                           unsigned int n_output_grains) const
+      {
+        // Get volume weighted euler angles, using random draws to convert odf
+        // to a discrete number of orientations, weighted by volume
+        // 1a. Get index that would sort volume fractions AND
+        //ix = np.argsort(fv[q,:]);
+        // 1b. Get the sorted volume and angle arrays
+        std::vector<double> fv_to_sort = fv;
+        std::vector<double> fv_sorted = fv;
+        std::vector<Tensor<2,3>> matrices_sorted = matrices;
+
+        unsigned int n_grain = fv_to_sort.size();
+
+
+        /**
+         * ...
+         */
+        for (int i = n_grain-1; i >= 0; --i)
+          {
+            unsigned int index_max_fv = std::distance(fv_to_sort.begin(),max_element(fv_to_sort.begin(), fv_to_sort.end()));
+
+            fv_sorted[i] = fv_to_sort[index_max_fv];
+            matrices_sorted[i] = matrices[index_max_fv];
+            fv_to_sort[index_max_fv] = -1;
+          }
+
+        // 2. Get cumulative weight for volume fraction
+        std::vector<double> cum_weight(n_grains);
+        std::partial_sum(fv_sorted.begin(),fv_sorted.end(),cum_weight.begin());
+        // 3. Generate random indices
+        std::uniform_real_distribution<> dist(0, 1);
+        std::vector<double> idxgrain(n_output_grains);
+        for (unsigned int grain_i = 0; grain_i < n_output_grains; ++grain_i)
+          {
+            idxgrain[grain_i] = dist(this->random_number_generator);
+          }
+
+        // 4. Find the maximum cum_weight that is less than the random value.
+        // the euler angle index is +1. For example, if the idxGrain(g) < cumWeight(1),
+        // the index should be 1 not zero)
+        std::vector<Tensor<2,3>> matrices_out(n_output_grains);
+        for (unsigned int grain_i = 0; grain_i < n_output_grains; ++grain_i)
```

Comment:
Check https://github.com/dealii/dealii/blob/38356ce009ba1b041bc97655d134bccd2480d8d1/source/particles/generators.cc#L389, this should make this algorithm much simpler (and NlogN instead of N^2).